rqd/rqd/rqcore.py

```diff
@@ -51,1061 +52,1211 @@
+            self.machine.reboot()
+        else:
+            log.warning("Shutting down RQD by request. pid(%s)", os.getpid())
+        self.network.stopGrpc()
+        # Using sys.exit would raise SystemExit, giving exception handlers a chance
+        # to block this
+        # pylint: disable=protected-access
+        os._exit(0)
 
-    def runWindows(self):
-        """The steps required to handle a frame under windows"""
-        frameInfo = self.frameInfo
-        runFrame = self.runFrame
+    def handleExit(self, signalnum, flag):
+        """Shutdown threads and exit RQD."""
+        del signalnum
+        del flag
+        self.shutdown()
 
-        self.__createEnvVariables()
-        self.__writeHeader()
+    def launchFrame(self, runFrame):
+        """This will setup for the launch the frame specified in the arguments.
+        If a problem is encountered, a CueException will be thrown.
+        @type   runFrame: RunFrame
+        @param  runFrame: rqd_pb2.RunFrame"""
+        log.info("Running command %s for %s", runFrame.command, runFrame.frame_id)
+        log.debug(runFrame)
 
-        try:
-            runFrame.command = runFrame.command.replace('%{frame}', self.frameEnv['CUE_IFRAME'])
-            tempCommand = [self._createCommandFile(runFrame.command)]
+        #
+        # Check for reasons to abort launch
+        #
 
-            # pylint: disable=consider-using-with
-            frameInfo.forkedCommand = subprocess.Popen(tempCommand,
-                                                       env=self.frameEnv,
-                                                       stdin=subprocess.PIPE,
-                                                       stdout=subprocess.PIPE,
-                                                       stderr=subprocess.STDOUT)
-        # pylint: disable=broad-except
-        except Exception:
-            log.critical(
-                "Failed subprocess.Popen: Due to: \n%s",
-                ''.join(traceback.format_exception(*sys.exc_info())))
+        if self.machine.state != rqd.compiled_proto.host_pb2.UP:
+            err = "Not launching, rqd HardwareState is not Up"
+            log.info(err)
+            raise rqd.rqexceptions.CoreReservationFailureException(err)
 
-        frameInfo.pid = frameInfo.forkedCommand.pid
+        if self.__whenIdle:
+            err = "Not launching, rqd is waiting for idle to shutdown"
+            log.info(err)
+            raise rqd.rqexceptions.CoreReservationFailureException(err)
 
-        if not self.rqCore.updateRssThread.is_alive():
-            self.rqCore.updateRssThread = threading.Timer(rqd.rqconstants.RSS_UPDATE_INTERVAL,
-                                                          self.rqCore.updateRss)
-            self.rqCore.updateRssThread.start()
+        if self.nimby.locked and not runFrame.ignore_nimby:
+            err = "Not launching, rqd is lockNimby and not Ignore Nimby"
+            log.info(err)
```

Comment:
I recommend rewriting to:

`Not launching, rqd is in the status lockNimby and option Ignore Nimby is not activated`